File: python_services/shared/src/shared_handler/kv.py
from nats.js.kv import KeyValue
from nats.js import JetStreamContext
from nats.js.api import KeyValueConfig
from nats.js.errors import KeyNotFoundError
from shared_core import get_logger, settings
import json
import nats.js.errors as js_errors


async def connect_kv(js: JetStreamContext, kv_name: str) -> KeyValue:
    """
    Connect to an existing jetstream kv

    Args:
        js: jetstreamContext server we are connecting to
        kv_name: the kv we are trying to connect to

    Returns:
        the Jetstream KeyValue connection

    Raises:
        RuntimeError if the Jetstream KV isnt found
    """
    try:
        job_status_kv = await js.key_value(kv_name)

        return job_status_kv
    except js_errors.NotFoundError:
        raise RuntimeError(
            f"{kv_name} KV bucket not found, check video-status is running"
        )


async def create_kv(js: JetStreamContext, bucket_name: str) -> KeyValue:
    """
    Create a new Jetstream KV

    Args:
        js: jetstreamContext server we creating the new KV on
        kv_name: the kv we are trying to create

    Returns:
        the Jetstream KeyValue connection

    Raises:
        RuntimeError if the a API error happens with jetstream
    """
    try:
        msg_processed_kv = await js.create_key_value(
            config=KeyValueConfig(bucket=bucket_name, ttl=settings.KV_BUCKET_TTL_S)
        )

        return msg_processed_kv

    except js_errors.APIError as e:
        raise RuntimeError(f"failed to create {bucket_name} KV bucket: {e}")


async def check_already_processed(kv: KeyValue, job_id: str) -> bool:
    """Checks if the job_id exists in the kv so it doesnt reprocess"""
    try:
        await kv.get(job_id)
        return True

    except KeyNotFoundError:
        return False


async def is_job_cancelled(job_milestone_kv: KeyValue, job_id: str) -> bool:
    """Check the job_milestone_kv for the state of the job_id
    and returns true if the value state is CANCELLED, false otherwise"""
    try:
        entry = await job_milestone_kv.get(job_id)
    except KeyNotFoundError:
        return False
    current = json.loads(entry.value) if entry.value is not None else {}

    return current.get("state") == "CANCELLED"


TERMINAL_MILESTONE_STATES = {"COMPLETE", "FAILED", "CANCELLED"}

MILESTONE_STAGE_ORDER = {
    "scene-detector": 0,
    "video-upscaling": 0,
    "video-recombiner": 1,
}


async def advance_milestone(
    job_milestone_kv: KeyValue,
    job_id: str,
    payload: dict[str, str],
) -> None:
    """
    Writes payload to the job-milestones KV bucket only if it isn't behind the
    currently stored stage and the job isn't already in terminal state
    """
    new_state = payload.get("state", "")
    new_stage = payload.get("stage", "")
    new_ordinal = MILESTONE_STAGE_ORDER.get(new_stage, -1)
    status = json.dumps(payload).encode()

    while True:
        entry = await job_milestone_kv.get(job_id)

        current = json.loads(entry.value) if entry.value is not None else {}
        current_state = current.get("state", "")
        current_stage = current.get("stage", "")
        current_ordinal = MILESTONE_STAGE_ORDER.get(current_stage, -1)

        if current_state in TERMINAL_MILESTONE_STATES:
            return
        if (
            new_state not in TERMINAL_MILESTONE_STATES
            and current_ordinal >= new_ordinal
        ):
            return

        try:
            await job_milestone_kv.update(job_id, status, last=entry.revision)
        except js_errors.KeyWrongLastSequenceError:
            continue  # revision changed concurrently; reread and compare again
        return


async def update_job_stage(
    job_milestone_kv: KeyValue,
    job_id: str,
    stage: str,
    service_name: str,
) -> None:
    """
    Writes the current processing stage to the job-milestones KV bucket

    Args:

    Exception:
        logs the error
    """
    logger = get_logger(service_name)

    try:
        payload: dict[str, str] = {"state": "PROCESSING", "stage": stage}
        await advance_milestone(job_milestone_kv, job_id, payload)
    except Exception as e:
        logger.error("failed to update job-milestones stage", job_id=job_id, err=str(e))
        raise


async def update_job_failed(
    job_milestone_kv: KeyValue,
    job_id: str,
    error: str,
    service_name: str,
) -> None:
    """
    Writes FAILED with the underlying error message to the job-milestones KV bucket

    Raises:
        Exception: re-raised if the KV write fails, after logging
    """
    logger = get_logger(service_name)

    payload = {"state": "FAILED", "error": error}
    try:
        await advance_milestone(job_milestone_kv, job_id, payload)
    except Exception as e:
        logger.error(
            "failed to update job-milestones to failed", job_id=job_id, err=str(e)
        )
        raise

File: go_services/internal/shared/jetstream/kv.go
package jetstream

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"time"

	"github.com/nats-io/nats.go/jetstream"
)

var osExit = os.Exit

// create the jetstream KV bucket for use in application
func CreateKV(
	bucketName string, js jetstream.JetStream, ttl time.Duration, logger *slog.Logger,
) jetstream.KeyValue {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	kv, err := js.CreateOrUpdateKeyValue(ctx, jetstream.KeyValueConfig{
		Bucket: bucketName,
		TTL:    ttl,
	})
	if err != nil {
		logger.Error("failed to create kv bucket", "bucketName", bucketName, "err", err)
		osExit(1)
		return nil
	}

	return kv
}

// connect to existing kv
func ConnectKV(js jetstream.JetStream, bucketName string, logger *slog.Logger) jetstream.KeyValue {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	kv, err := js.KeyValue(ctx, bucketName)
	if err != nil {
		logger.Error("failed to connect to kv bucket", "bucketName", bucketName, "err", err)
		osExit(1)
	}

	return kv
}

// check if a key exists in the KV
func CheckKeyExist(kv jetstream.KeyValue, key string) (bool, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	_, err := kv.Get(ctx, key)
	if err != nil {
		if errors.Is(err, jetstream.ErrKeyNotFound) {
			return false, nil
		}
		return false, fmt.Errorf("failed: %w", err)
	}

	return true, nil
}

// puts a new key value pair into the jetstream kv. Note the context is hardc
func PutKeyKV(kv jetstream.KeyValue, key string, value []byte) error {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	_, err := kv.Put(ctx, key, value)
	if err != nil {
		return fmt.Errorf("failed: %w", err)
	}

	return nil
}

// reads a job's milestone entry for the jobID and returns the revision, JobStatus, and err (if applicable)
func GetMilestoneKV(kv jetstream.KeyValue, jobID string) (uint64, JobStatus, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	entry, err := kv.Get(ctx, jobID)
	if errors.Is(err, jetstream.ErrKeyNotFound) {
		return 0, JobStatus{}, nil
	}
	if err != nil {
		return 0, JobStatus{}, fmt.Errorf("failed to fetch from kv: %w", err)
	}

	var current JobStatus
	err = json.Unmarshal(entry.Value(), &current)
	if err != nil {
		return 0, JobStatus{}, fmt.Errorf("failed to unmarshal json: %w", err)
	}

	return entry.Revision(), current, nil
}

// For job mileston kv

// ranks pipeline stages so AdvanceMilestone can differentiate forward write from stale one
var milestoneStageOrder = map[string]int{
	"upload":           0,
	"transcoder":       1,
	"video-recombiner": 2,
}

// writes a job's milestone entry only if newStage is not behind currently stored stage and
// job isnt already in terminal state.
func AdvanceMilestone(kv jetstream.KeyValue, jobID string, newStatus JobStatus) error {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	newValue, err := json.Marshal(newStatus)
	if err != nil {
		return fmt.Errorf("failed: %w", err)
	}

	for {
		revision, jobStatus, err := GetMilestoneKV(kv, jobID)
		if err != nil {
			return fmt.Errorf("failed: %w", err)
		}
		if jobStatus.State == "" {
			_, err = kv.Create(ctx, jobID, newValue)
			if errors.Is(err, jetstream.ErrKeyExists) {
				continue // lost create race, reread and compare against winner
			}
			if err != nil {
				return fmt.Errorf("failed: %w", err)
			}
			return nil
		}

		if IsStaleTransition(jobStatus, newStatus) {
			return nil
		}

		_, err = kv.Update(ctx, jobID, newValue, revision)
		if errors.Is(err, jetstream.ErrKeyExists) {
			continue // revision change concurrently, reread and compare again
		}
		if err != nil {
			return fmt.Errorf("failed: %w", err)
		}
		return nil
	}
}

// For video/file chunks

// atomically claims jobID/chunkIndex via claimKV and if claim succeeds runs fn
// fn should return completed=true when work is durable persisted and false otherwise
// so legitimate retry can reclaim the chunk. On success claim is left in place to
// expire via claimKV TTL rather than released immediately
func ClaimAndRun(
	claimKV jetstream.KeyValue, jobID string, chunkIndex int, logger *slog.Logger, fn func() (completed bool),
) (bool, error) {
	claimed, err := claimChunk(claimKV, jobID, chunkIndex)
	if err != nil {
		return false, err
	}

	if !claimed {
		return false, nil
	}

	// track whether processing finished successfully so defer release chunk fires on failure path
	// this lets a legit retry reclaim the chunk. On success the claim is left to expire via claimKV TTL
	completed := false
	defer func() {
		if !completed {
			relErr := releaseChunkClaim(claimKV, jobID, chunkIndex)
			if relErr != nil {
				logger.Error("failed to release chunk claim", "job_id", jobID, "chunk_index", chunkIndex, "err", relErr)
			}
		}
	}()

	completed = fn()

	return true, nil
}

// ClaimChunk atomically claims a jobID/chunkIndex pair for processing. Returns claimed=false
// (no error) if another worker already holds the claim, so the caller can tell "I own this"
// apart from "someone else does" apart from a real KV failure.
func claimChunk(claimKV jetstream.KeyValue, jobID string, chunkIndex int) (bool, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	_, err := claimKV.Create(ctx, fmt.Sprintf("%s.%d", jobID, chunkIndex), []byte("in-progress"))
	if err != nil {
		if errors.Is(err, jetstream.ErrKeyExists) {
			return false, nil
		}
		return false, fmt.Errorf("failed: %w", err)
	}

	return true, nil
}

// ReleaseChunkClaim releases a chunk claim so a legitimate retry can reclaim it.
func releaseChunkClaim(claimKV jetstream.KeyValue, jobID string, chunkIndex int) error {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	err := claimKV.Delete(ctx, fmt.Sprintf("%s.%d", jobID, chunkIndex))
	if err != nil {
		return fmt.Errorf("failed: %w", err)
	}

	return nil
}
